File: CamelCase Pattern Matching.java
class Trie
{
    Trie arr[] = new Trie[26];
    boolean isComp = false; 
}

class Solution{
    
    Trie root = new Trie();
    
    public void makeTrie(String str)
    {
        Trie temp = root;
        
        for(char x : str.toCharArray())
        {
            Trie node = new Trie();
            int ind = x-'0'-17;
            temp.arr[ind] = node;
            temp = temp.arr[ind];
        }
        temp.isComp = true;
    }
    
    public boolean searchTrie(String str)
    {
        Trie temp = root;
        for(char x : str.toCharArray())
        {

            if(x<'A' || x>'Z')
                 continue;
                 
            int ind = x-'0'-17;
            temp = temp.arr[ind];

            if(temp == null)
                break;
            
            if(temp.isComp) return true;
        }
        return false;
    }
    
    ArrayList<String> CamelCase(int N,String[] Dictionary,String Pattern){
        
        ArrayList<String> ans = new ArrayList<>();
        
        makeTrie(Pattern);
        
        for(String str :Dictionary)
        {
            if(searchTrie(str))
                ans.add(str);
        }
        
        if(ans.size() == 0)
            ans.add("-1");
        
        Collections.sort(ans);
        
        return ans;
    }
}

File: CamelCase Pattern Matching.cpp
class Solution {
  public:
    vector<string> CamelCase(int N, vector<string> d, string Pattern) {
        vector<string> ans;
        string cmp;
        for(int j = 0; j < N; j++){
            string str = d[j];
            cmp = "";
            for(int i=0;i<str.size();i++){
                if (str[i] >= 'A' && str[i] <= 'Z') cmp += str[i];
                if (cmp.size() == Pattern.size()) break;
            }
            if (Pattern == cmp) ans.push_back(str);
        }
        if (ans.size() == 0) ans.push_back("-1");
        return ans;
    }
};
